guppylang/diagnostic.py

```diff
@@ -109,3 +151,222 @@ class Error(Diagnostic, Protocol):
+    ) -> None:
+        """Renders the source associated with a span together with an optional label.
+
+        ```
+           |
+        42 | def foo(x: blah) -> None:
+           |            ^^^^ Span label. This could cover
+           |                 multiple lines!
+        ```
+
+        Also supports spans covering multiple lines:
+
+        ```
+           |
+        42 | def foo(x: int) -> None:
+           | ^^^^^^^^^^^^^^^^^^^^^^^^
+           | ...
+        48 |     return bar()
+           | ^^^^^^^^^^^^^^^^ Label covering the entire definition of foo
+        ```
+
+        If `is_primary` is `False`, the span is highlighted using `-` instead of `^`:
+
+        ```
+           |
+        42 | def foo(x: blah) -> None:
+           |            ---- Non-primary span label
+        ```
+
+        Optionally includes up to `prefix_lines` preceding source lines to give
+        additional context.
+        """
+        # Check how much space we need to reserve for the leading line numbers
+        ll_length = len(str(span.end.line))
+        highlight_char = "^" if is_primary else "-"
+
+        def render_line(line: str, line_number: int | None = None) -> None:
+            """Helper method to render a line with the line number bar on the left."""
+            if line_number is not None:
+                ll = str(line_number)
+                self.buffer.append(ll + " " * (ll_length - len(ll)) + " | " + line)
+            else:
+                self.buffer.append(" " * ll_length + " | " + line)
+
+        # One line of padding
+        render_line("")
+
+        # Grab all lines we want to display and remove excessive leading whitespace
+        prefix_lines = min(prefix_lines, span.start.line - 1)
+        all_lines = self.source.span_lines(span, prefix_lines)
+        leading_whitespace = min(len(line) - len(line.lstrip()) for line in all_lines)
+        if leading_whitespace > self.MAX_LEADING_WHITESPACE:
+            remove = leading_whitespace - self.OPTIMAL_LEADING_WHITESPACE
+            all_lines = [line[remove:] for line in all_lines]
+            span = span.shift_left(remove)
+
+        # Render prefix lines
+        for i, line in enumerate(all_lines[:prefix_lines]):
+            render_line(line, span.start.line - prefix_lines + i)
+        span_lines = all_lines[prefix_lines:]
+
+        if span.is_multiline:
+            [first, *middle, last] = span_lines
+            render_line(first, span.start.line)
+            # Compute the subspan that only covers the first line and render it's
+            # highlight banner
+            first_span = Span(span.start, Loc(span.file, span.start.line, len(first)))
+            first_highlight = " " * first_span.start.column + highlight_char * len(
+                first_span
+            )
+            render_line(first_highlight)
+            # Omit everything in the middle
+            if middle:
```

Comment:
`if len(middle)>1`, maybe. It'd be silly to omit a single line to display a line `...`

Reply:
Oh, but I guess then you'd need to highlight that line as well, hmmm. Ok. Not sure about this, maybe I should just shut up and delete the parent comment ;)